trie/committer.go

```diff
@@ -139,15 +139,11 @@ func (c *committer) store(path []byte, n node) node {
 		}
 		return n
 	}
-	// We have the hash already, estimate the RLP encoding-size of the node.
-	// The size is used for mem tracking, does not need to be exact
 	var (
-		size  = estimateSize(n)
 		nhash = common.BytesToHash(hash)
 		mnode = &memoryNode{
 			hash: nhash,
-			node: simplifyNode(n),
-			size: uint16(size),
+			node: nodeToBytes(n),
```

Comment:
I think about it. After the change in this PR, there are some extra costs for verkle, mostly the node serialization and deserialization:

#### Scenario description

There are a few hundreds accounts and a few thousands touched in the mainnet block. Let's use 500 accounts and 3.5K slots here for performance estimation.

In verkle, the node width is 256, the average tree depth is 4. If we need to touch 4K leaves in total, it means 4K leaf nodes and 585(1+8+8^2+8^3) internal nodes will be accessed.

#### Deserialization
It happens when the node is accessed. The major cost is the conversion from node hash to node commitment(find the elliptic curve point by its X coordinator). I wrote a benchmark and it turns out it's indeed extremely expensive.

```go
// BenchmarkHashToCommitment
// BenchmarkHashToCommitment-8   	   45598	     25769 ns/op	      32 B/op	       1 allocs/o
func BenchmarkHashToCommitment(b *testing.B) {
	p := &Generator
	p.Add(p, p)
	hash := p.Bytes()

	b.ResetTimer()
	b.ReportAllocs()

	for i := 0; i < b.N; i++ {
		var elem Element
		elem.SetBytesTrusted(hash[:])
	}
}
```

If we need to do the conversion for 4.58K nodes, the overall cost will be 112ms. Fortunately we have trie prefetcher which will try to load and decode nodes during evm execution. Not sure how much can be covered by it though.

#### Serialization

The same, the major cost of serialization comes from the conversion of node commitment to node hash. And it's possible to speed it up by running the conversion in a batch.

So the total cost of this conversion is 4.5ms and might be possible to reduce it a lot.

```go
// BenchmarkCommitmentToHash
// BenchmarkCommitmentToHash-8   	  740104	      1625 ns/op	       0 B/op	       0 allocs/op
func BenchmarkCommitmentToHash(b *testing.B) {
	p := &Generator
	p.Add(p, p)

	b.ResetTimer()
	b.ReportAllocs()

	for i := 0; i < b.N; i++ {
		p.Bytes()
	}
}

// nodes = 16, 384.625 ns in average
// BenchmarkElementsToBytes
// BenchmarkElementsToBytes-8   	  169038	      6154 ns/op	    1552 B/op	       4 allocs/op
func BenchmarkElementsToBytes16(b *testing.B) {
	benchmarkElementsToBytes(b, 16)
}

// nodes = 256, 250.41015625 ns in average
// BenchmarkElementsToBytes256
// BenchmarkElementsToBytes256-8   	   17974	     64105 ns/op	   24832 B/op	       4 allocs/op
func BenchmarkElementsToBytes256(b *testing.B) {
	benchmarkElementsToBytes(b, 256)
}
```

#### Summary

The cost of node serialization and deserialization is not trivial. But if we keep the node or a pointer of node in memory, it will require some additional abstraction and might be MPT or Verkle specific. I hope we can optimize a bit to reduce the cost.

Reply:
But as you said, if we change the serialization rules a bit, by retaining the commitment(elliptic curve point) itself instead of the hash(x coordination of point), then we can get rid of this part of cost. But the cost is node size is larger.

I think we really need to twist the trie storage scheme, since currently the verkle node can have at most 256 children and the size of child is 32B. The database amplification is already very large. We need to somehow split the children set to reduce the database amplification.